plugin/semisync/semisync_master_ack_receiver.cc

```diff
@@ -203,6 +214,91 @@ void Ack_receiver::remove_slave(THD *thd) {
   function_exit(kWho);
 }
 
+void Ack_receiver::disconnect_non_whitelisted_slaves() {
+  const char *kWho = "Ack_receiver::remove_not_whitelisted_slaves";
+  function_enter(kWho);
+
+  mysql_mutex_assert_owner(&m_mutex);
+
+  if (rpl_semi_sync_master_whitelist &&
+      strcmp(rpl_semi_sync_master_whitelist, any_slave) == 0) {
+    return function_exit(kWho);
+  }
+
+  for (Slave_vector_it it = m_slaves.begin(); it != m_slaves.end(); ++it) {
+    if (!rpl_semi_sync_master_whitelist ||
+        !verify_against_whitelist(it->slave_uuid))
+      vio_shutdown(it->vio);
```

Comment:
Since this is happening on the connection executing the SET statement, it seems like this is not thread safe. Similar calls to vio_shutdown() in sql_class.cc THD::shutdown_active_vio asserts the thd lock is held. Do we need to stash the thd pointer so that the thread can be killed vs. calling vio_shutdown() directly?

Reply:
It seems that:
1. Only `THD::active_vio` field is protected by `LOCK_thd_data` (look at `THD::set_active_vio`)
2. `Ack_receiver::run` uses `net.vio` (e.g. `my_net_read(&net)`) without any additional mutex
3. `struct Vio` can be thread-safe with `std::atomic_flag poll_shutdown_flag`

Reply:
This is probably fine then. The read code doesn't acquire mutex, so killing the thread probably shuts down the vio in a similar way. I wasn't sure if the perf schema access code prior to poll_shutdown_flag is thread-safe, but looking through it, it seems like it should be. In any case, it's probably good enough.